### PR TITLE
streamingccl: switch to CREATE TENANT ... FROM REPLICATION

### LINKED ISSUE #91240: c2c: Extend `CREATE TENANT` to support `FROM REPLICATION`
Once we have #91235 we should expand the grammar to support creating the tenant from a replication stream.  The complete grammar should be:

`CREATE TENANT <destination-name> FROM REPLICATION OF <source-name> ON <connection-strings>`

This will replace the existing `RESTORE TENANT $1 FROM REPLICATION STREAM FROM string_or_placeholder_opt_list opt_as_tenant_clause` rule that we have today. Note, as part of this work we will want to persist the ingestion job ID in the tenant record. This coupling is necessary to know whether a tenant is being streamed into or not, and for future interactions such as `DROP` and `ALTER` that will use the tenant to resolve the ingestion job. 

Jira issue: CRDB-21170

Epic CRDB-18749

### PR DESCRIPTION
**Pull request by adityamaru**
This change is a refactor of the replication logic
to support the new syntax that addresses tenants by name
instead of ID.

When setting up the stream ingestion job on the destination
cluster we do not know the source tenant ID. This information is necessary
for us to setup our rekeyer that is used to rekey every KV streamed
from the source before ingestion. The `SourceTenantID` is obtained from
the source cluster as part of the `Topology` that governs how we
setup our streaming partitions.

The rest of the streaming logic is largely unchanged.

Fixes: https://github.com/cockroachdb/cockroach/issues/91240

Release note: None

### CODE REVIEW

**Comment by adityamaru** [comment]
I want CI to shake out stuff so please hold off on reviewing this PR for now!

**Review by lidorcarmel — CHANGES_REQUESTED** [review]
I did a quick pass.. just a few questions.

> Reviewed 5 of 5 files at r1, 36 of 36 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/92070#-:-NH6lGEjB3ourEKRtVIh:b-jj6284)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @baoalvin1, @dt, and @stevendanna)

___
*[`-- commits` line 12 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/92070#-NH6e3dg8TQwgrsYZ-Vs):*
did we say no release notes? so that we will not have docs for now?



_Code quote:_
```
  Release note (sql change): add syntax to create a tenant whose data will be
  streamed from a tenant on the source cluster
```
___
*[`pkg/ccl/streamingccl/streamclient/client.go` line 119 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/92070#-NH6l9DeDpH17v4v50y-:-NH6l9DeDpH17v4v50y0:b-d8xq1s) ([raw file](https://github.com/cockroachdb/cockroach/blob/4df6eceb4219f23e5ee077ee21deb37e32e2ef6a/pkg/ccl/streamingccl/streamclient/client.go#L119)):*
> ```Go
> 	SrcAddr       streamingccl.PartitionAddress
> 	SrcLocality   roachpb.Locality
> 	SrcTenantID   roachpb.TenantID
> ```

can/should we use TenantName here instead of ID?



_Code quote:_
```Go
SrcTenantID   roachpb.TenantID
```
___
*[`pkg/sql/sem/tree/create.go` line 2190 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/92070#-NH6eiCR4gmMjBrLeDs1:-NH6eiCR4gmMjBrLeDs2:bowy7p5) ([raw file](https://github.com/cockroachdb/cockroach/blob/596777c5b59b74c9a4ab5f2631721e020959a6aa/pkg/sql/sem/tree/create.go#L2190)):*
> ```Go
> 	// ReplicationSourceAddress is the address of the source cluster that we are
> 	// replicating data from.
> 	ReplicationSourceAddress Expr
> ```

those probably are not needed here?

can we drop `CreateTenant` completely and have only `CreateTenantFromReplication`? (and rename CreateTenantFromReplication to CreateTenant... I think you see where I'm going..)



_Code quote:_
```Go
	// ReplicationSourceTenantName is the name of the tenant that we are
	// replicating into the newly created tenant.
	ReplicationSourceTenantName Name
	// ReplicationSourceAddress is the address of the source cluster that we are
	// replicating data from.
	ReplicationSourceAddress Expr
```

File: pkg/sql/parser/sql.y [hunk 1]
```diff
@@ -4163,6 +4163,14 @@ create_tenant_stmt:
   {
     $$.val = &tree.CreateTenant{Name: tree.Name($3)}
   }
+| CREATE TENANT name FROM REPLICATION OF name ON string_or_placeholder
```
Comment (stevendanna) [inline pkg/sql/parser/sql.y:4166]: I think this statement might be the one where we really end up wanting to use placeholders for these names.  But, no need to change it now, we can do a follow-up PR where we just change them.

**Review by stevendanna — COMMENTED** [review]
Nice. Overall looks reasonable.

I think there is at least 1 more use of the old syntax over in the roachtest.

In a follow up PR, we probably want to rip out the `RESTORE TENANT ... FROM REPLICATION` syntax from sql.y and friends as well.

**Review by adityamaru — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/92070)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @baoalvin1, @dt, @lidorcarmel, and @stevendanna)

___
*[`-- commits` line 12 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/92070#-NH6e3dg8TQwgrsYZ-Vs):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

did we say no release notes? so that we will not have docs for now?
</blockquote></details>

Old habits die hard, removing!
___
*[`pkg/ccl/streamingccl/streamclient/client.go` line 119 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/92070#-NH6l9DeDpH17v4v50y-:-NHA5omR6oD6eBH1QU8h:b-w4c0qx) ([raw file](https://github.com/cockroachdb/cockroach/blob/4df6eceb4219f23e5ee077ee21deb37e32e2ef6a/pkg/ccl/streamingccl/streamclient/client.go#L119)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

can/should we use TenantName here instead of ID?
</blockquote></details>

nice catch, its now an unused field. Deleted.
___
*[`pkg/ccl/streamingccl/streamproducer/stream_lifetime.go` line 35 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/92070#-NH5LLgMhsaUQHKV6Qkp-r2-35:-NHA62IS4lTdn1C--g47:b-usram9) ([raw file](https://github.com/cockroachdb/cockroach/blob/4df6eceb4219f23e5ee077ee21deb37e32e2ef6a/pkg/ccl/streamingccl/streamproducer/stream_lifetime.go#L35)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Looking at this file, it looks like this is done? We create the PTS record in this function, advanced in updateReplicationStreamProgress, and then released in the producer job's OnFailOrCancel routine.
</blockquote></details>

indeed, removed!
___
*[`pkg/sql/parser/sql.y` line 4166 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/92070#-NH5LLgLuLEZuf-mthw_-r2-4166:-NHA65JsEpYNDk35YiGo:b-zab4g0) ([raw file](https://github.com/cockroachdb/cockroach/blob/4df6eceb4219f23e5ee077ee21deb37e32e2ef6a/pkg/sql/parser/sql.y#L4166)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

I think this statement might be the one where we really end up wanting to use placeholders for these names.  But, no need to change it now, we can do a follow-up PR where we just change them.
</blockquote></details>

:thumbsup: makes sense

**Comment by adityamaru** [comment]
> I think there is at least 1 more use of the old syntax over in the roachtest.

@stevendanna i tacked on one more commit with changes to the roachtest. I'm just running it to confirm I didn't break anything but it looks to be running smoothly as of now.

**Review by adityamaru — COMMENTED** [review]
Yup, I'll fast follow with the deletion. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/92070#-:-NHA7e_xDLp-50cTUA1_:bce9ztr)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @baoalvin1, @dt, @lidorcarmel, and @stevendanna)

___
*[`pkg/sql/sem/tree/create.go` line 2190 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/92070#-NH6eiCR4gmMjBrLeDs1:-NHA6PxyDnVxOCU6hvtI:bues4c4) ([raw file](https://github.com/cockroachdb/cockroach/blob/596777c5b59b74c9a4ab5f2631721e020959a6aa/pkg/sql/sem/tree/create.go#L2190)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

those probably are not needed here?

can we drop `CreateTenant` completely and have only `CreateTenantFromReplication`? (and rename CreateTenantFromReplication to CreateTenant... I think you see where I'm going..)
</blockquote></details>

my bad, this is detritus from when I was debating between introducing a new tree.Node or re-using `CreateTenant`. 

The reason we need two distinct nodes is that `CreateTenantFromReplication` is a CCL node which means it relies on having a `planHook` that is used to plan the statement (see stream_ingestion_planning.go`). This is the pattern we follow for calling CCL code from a non-CCL package like `pkg/sql` since having a non-CCL package import a CCL package is not allowed except for testing purposes. `CreateTenant` OTOH is a non-ccl node and doesn't require a planhook for execution.

**Review by lidorcarmel — APPROVED** [review]
> Reviewed 38 of 38 files at r3, 37 of 37 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/92070)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @baoalvin1, @dt, and @stevendanna)

___
*[`pkg/cmd/roachtest/tests/cluster_to_cluster.go` line 109 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/92070#-NHAvVhm-yeuV1GTyZFz:-NHAvVhnBhj1aoPvJ36z:b-e8fl1i) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd00c9564bdd0dd1f0f44df116c77dd7bf50cb23/pkg/cmd/roachtest/tests/cluster_to_cluster.go#L109)):*
> ```Go
> 
> 			t.Status("starting replication stream")
> 			streamReplStmt := fmt.Sprintf("CREATE TENANT \"%s\" FROM REPLICATION OF \"%s\" ON '%s'",
> ```

tiny nit: maybe use `%q` to avoid the `\"`? or quote everything with `'` instead of `"`?



_Code quote:_
```Go
\"%s\" FROM REPLICATION OF \"%s\" 
```
___
*[`pkg/sql/sem/tree/create.go` line 2190 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/92070#-NH6eiCR4gmMjBrLeDs1:-NHAvp2GBnO4JGqocIHx:b-n1918y) ([raw file](https://github.com/cockroachdb/cockroach/blob/596777c5b59b74c9a4ab5f2631721e020959a6aa/pkg/sql/sem/tree/create.go#L2190)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

my bad, this is detritus from when I was debating between introducing a new tree.Node or re-using `CreateTenant`. 

The reason we need two distinct nodes is that `CreateTenantFromReplication` is a CCL node which means it relies on having a `planHook` that is used to plan the statement (see stream_ingestion_planning.go`). This is the pattern we follow for calling CCL code from a non-CCL package like `pkg/sql` since having a non-CCL package import a CCL package is not allowed except for testing purposes. `CreateTenant` OTOH is a non-ccl node and doesn't require a planhook for execution.
</blockquote></details>

good to know, thanks for explaining!

**Review by adityamaru — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/92070)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @baoalvin1, @dt, @lidorcarmel, and @stevendanna)

___
*[`pkg/cmd/roachtest/tests/cluster_to_cluster.go` line 109 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/92070#-NHAvVhm-yeuV1GTyZFz:-NHBR0pp-BCFdHOvRrje:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd00c9564bdd0dd1f0f44df116c77dd7bf50cb23/pkg/cmd/roachtest/tests/cluster_to_cluster.go#L109)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

tiny nit: maybe use `%q` to avoid the `\"`? or quote everything with `'` instead of `"`?
</blockquote></details>

Done.

**Review by stevendanna — APPROVED** [review]

**Comment by adityamaru** [comment]
TFTRs!

bors r=lidorcarmel,stevendanna

**Pull request merged**